pipeline/beam_tables.py

```diff
@@ -425,6 +464,70 @@ def _write_to_bigquery(self, scan_type: str,
         write_disposition=write_mode,
         additional_bq_parameters=_get_partition_params()))
 
+  def _write_to_gcs(  # pylint: disable=no-self-use
+      self, scan_type: str, rows: beam.pvalue.PCollection[BigqueryRow],
+      gcs_folder: str) -> None:
+    """Write out rows to GCS folder with hive-partioned format.
+
+    Rows are written to .json.gz files organized by source and country:
+    '{gcs_folder}/source={source}/country={country}/results.json.gz'
+
+    Args:
+      scan_type: one of 'echo', 'discard', 'http', 'https',
+        or 'satellite'
+      rows: PCollection[BigqueryRow] of data to write.
+      gcs_folder: GCS folder gs://bucket/folder/... like
+        'gs://firehook-test/scans/echo' to write output files to.
+
+    Raises:
+      Exception: if any arguments are invalid.
+    """
+
+    def _get_destination(record: str) -> str:
+      """Returns the hive-format dest folder for a measurement record str."""
+      record_dict = json.loads(record)
+      return f'source={record_dict["source"]}/country={record_dict["country"]}/results'
+
+    def _custom_file_naming(suffix: str = None) -> Callable:
+      """Returns custom function to name destination files."""
+
+      # Beam requires the returned function to have the following arguments,
+      # see beam.io.filename.destination_prefix_naming.
+      def _inner(window: Any, pane: Any, shard_index: Optional[int],
+                 total_shards: Optional[int], compression: Optional[str],
+                 destination: Optional[str]) -> str:
+        # Get the filename with the destination_prefix_naming format:
+        # '{prefix}-{start}-{end}-{pane}-{shard:05d}-of-{total_shards:05d}{suffix}{compression}'
+        # where prefix = destination.
+        filename = beam.io.fileio.destination_prefix_naming(suffix)(
+            window, pane, shard_index, total_shards, compression, destination)
+        # Remove shard component from filename if there is only one shard.
+        filename = filename.replace('-00000-of-00001', '')
+        logging.info('%s %s %s %s', destination, shard_index, total_shards,
+                     filename)
+        return filename
+
+      return _inner
+
+    # Pcollection[Dict[str, Any]]
+    dict_rows = (
+        rows | f'dataclass to dicts {scan_type}' >> beam.Map(
+            schema.flatten_for_bigquery).with_output_types(Dict[str, Any]))
+
+    # PCollection[str]
+    json_rows = (
```

Comment:
If we do the comment above, this and the step above will become a map over a single method which would live in schema.py